lib/src/utils/parsing/parsing_utils.dart

```diff
@@ -0,0 +1,321 @@
+
+  // total collateral  
+  final totalCollateral = tx.totalCollateral == null
+    ? null  
+    : parseCoin(tx.totalCollateral, InvalidDataReason.totalCollateralNotValid);
+
+  // reference inputs
+  validate(isArray(tx.referenceInputs ?? []), InvalidDataReason.referenceInputsNotArray.message);
+  final referenceInputs = (tx.referenceInputs ?? []).map((ri) => parseTxInput(ri)).toList();
+
+  // voting procedures
+  validate(isArray(tx.votingProcedures ?? []), InvalidDataReason.votingProceduresNotArray.message);
+  final votingProcedures = (tx.votingProcedures ?? []).map((x) => parseVoterVotes(x)).toList();
+  validate(votingProcedures.length <= 1, InvalidDataReason.votingProceduresInvalidNumberOfVoters.message);
+  for (final voterVotes in votingProcedures) {
+    validate(voterVotes.votes.length == 1, InvalidDataReason.votingProceduresInvalidNumberOfVotes.message);
+  }
+
+  // treasury
+  final treasury = tx.treasury == null
+    ? null
+    : parseCoin(tx.treasury, InvalidDataReason.treasuryNotValid);
+
+  // donation
+  final donation = tx.donation == null
+    ? null
+    : parseUint64Str(tx.donation, {'min': '1', 'max': maxLovelaceSupplyStr}, InvalidDataReason.donationNotValid);
+
+  return ParsedTransaction(
+    network: network,
+    inputs: inputs,
+    outputs: outputs,
+    ttl: ttl == null ? null : BigInt.parse(ttl),
+    auxiliaryData: auxiliaryData,
+    validityIntervalStart: validityIntervalStart == null ? null : BigInt.parse(validityIntervalStart),
+    withdrawals: withdrawals,
+    certificates: certificates,
+    fee: BigInt.parse(fee),
+    mint: mint,
+    scriptDataHashHex: scriptDataHashHex,
+    collateralInputs: collateralInputs,
+    requiredSigners: requiredSigners,
+    includeNetworkId: includeNetworkId,
+    collateralOutput: collateralOutput,
+    totalCollateral: totalCollateral == null ? null : BigInt.parse(totalCollateral),
+    referenceInputs: referenceInputs,
+    votingProcedures: votingProcedures,
+    treasury: treasury == null ? null : BigInt.parse(treasury),
+    donation: donation == null ? null : BigInt.parse(donation),
+  );
+}
+
+ParsedNetwork parseNetwork(Network network) {
+  final parsed = ParsedNetwork(
+    protocolMagic: parseUint32t(
+      network.protocolMagic,
+      InvalidDataReason.networkInvalidProtocolMagic.message,
+    ),
+    networkId: parseUint8t(
+      network.networkId,
+      InvalidDataReason.networkInvalidNetworkId.message,
+    ),
+  );
+  validate(
+    parsed.networkId <= 0x0F,
+    InvalidDataReason.networkInvalidNetworkId.message,
+  );
+  return parsed;
+}
+
+int parseUint32t(dynamic value, String errMsg) {
+  validate(isUint32(value), errMsg);
+  return value as int;
+}
+
+
+bool isUint32(dynamic data) =>
+  isInteger(data) && data >= 0 && data <= 4294967295;
+
+bool isInteger(dynamic data) =>
+  data is int || data is BigInt;
+
+bool isUint8(dynamic data) =>
+  isInteger(data) && data >= 0 && data <= 255;
```

Comment:
You shouldn't use `dynamic` because you're adding bugs without even realizing. this for example is wrong here. if data is BigInt you're comparing a BigInt with int which will throw error. 

If you relied on smart casting and actual non-dynamic types it wouldn't have allowed you to do this in the first place.